### PR TITLE
Filter sensitive fields from debug output

### LINKED ISSUE #13: Hide passwords and tokens from the log
Moved from: https://github.com/openshift/autoheal/issues/43

These come from HTTP debugging of payload and headers. We should apply regular expressions to hide them. In the output below you can see 2 issues both the basic auth as well as the returning oAuth token:
```
I0411 12:41:53.363281   20313 connection.go:398] Sending POST request to 'http://localhost:9100/api/v2/authtoken/'.
I0411 12:41:53.363314   20313 connection.go:399] Request body:
{
  "username": "admin",
  "password": "password"
}
I0411 12:41:53.363344   20313 connection.go:400] Request headers:
I0411 12:41:53.363370   20313 connection.go:402] 	Content-Type: [application/json]
I0411 12:41:53.363400   20313 connection.go:402] 	Accept: [application/json]
I0411 12:41:53.363424   20313 connection.go:402] 	User-Agent: []
I0411 12:41:53.363449   20313 connection.go:402] 	Authorization: [Basic YWRtaW46cGFzc3dvcmQ=]
I0411 12:41:53.525526   20313 connection.go:418] Response body:
{
  "detail": "The requested resource could not be found."
}
I0411 12:41:53.525560   20313 connection.go:419] Response headers:
I0411 12:41:53.525568   20313 connection.go:421] 	Connection: [keep-alive]
I0411 12:41:53.525577   20313 connection.go:421] 	Vary: [Accept, Accept-Language, Cookie]
I0411 12:41:53.525584   20313 connection.go:421] 	X-Api-Total-Time: [0.006s]
I0411 12:41:53.525590   20313 connection.go:421] 	Content-Language: [en]
I0411 12:41:53.525597   20313 connection.go:421] 	Server: [nginx/1.12.2]
I0411 12:41:53.525604   20313 connection.go:421] 	Content-Type: [application/json]
I0411 12:41:53.525610   20313 connection.go:421] 	Content-Length: [55]
I0411 12:41:53.525617   20313 connection.go:421] 	Date: [Wed, 11 Apr 2018 09:41:53 GMT]
W0411 12:41:53.525626   20313 connection.go:229] Failed to aquire authtoken 'Status code '404' returned from server: '404 Not Found'', attempting PAT
I0411 12:41:53.526114   20313 connection.go:398] Sending POST request to 'http://localhost:9100/api/v2/users/admin/personal_tokens/'.
I0411 12:41:53.526128   20313 connection.go:399] Request body:
{
  "description": "AWX Go Client",
  "application": null,
  "scope": "write"
}
I0411 12:41:53.526138   20313 connection.go:400] Request headers:
I0411 12:41:53.526144   20313 connection.go:402] 	User-Agent: []
I0411 12:41:53.526152   20313 connection.go:402] 	Authorization: [Basic YWRtaW46cGFzc3dvcmQ=]
I0411 12:41:53.526159   20313 connection.go:402] 	Content-Type: [application/json]
I0411 12:41:53.526166   20313 connection.go:402] 	Accept: [application/json]
I0411 12:41:54.005212   20313 connection.go:418] Response body:
{
  "id": 1,
  "type": "o_auth2_access_token",
  "url": "/api/v2/tokens/1/",
  "related": {
    "user": "/api/v2/users/1/",
    "activity_stream": "/api/v2/tokens/1/activity_stream/"
  },
  "summary_fields": {
    "user": {
      "id": 1,
      "username": "admin",
      "first_name": "",
      "last_name": ""
    }
  },
  "created": "2018-04-11T09:41:53.732408Z",
  "modified": "2018-04-11T09:41:53.740174Z",
  "description": "AWX Go Client",
  "user": 1,
  "token": "DIfUPU5Br8hXbwb5BTTZOWiY8chVKJ",
  "refresh_token": null,
  "application": null,
  "expires": "3017-08-12T09:41:53.731847Z",
  "scope": "write"
}
```

### PR DESCRIPTION
**Pull request by elad661**
This PR attempts to hide passwords and tokens from the debug log.

Fixes #13

cc: @moolitayer @zgalor @nimrodshn @jhernand

### CODE REVIEW

File: awx/connection.go [hunk 1]
```diff
@@ -533,3 +534,49 @@ func (c *Connection) indent(data []byte) []byte {
 	}
 	return buffer.Bytes()
 }
+
+var passwordFilterRegex = regexp.MustCompile("(?i)(password|token|authorization|key)")
+
+func filterJsonBytes(bytes []byte) []byte {
+	if len(bytes) == 0 || bytes == nil {
```
Comment (moolitayer) [inline awx/connection.go]: nit, len(bytes) == 0 capture bytes == nil

**Comment by moolitayer** [comment]
@elad661 can you write a few simple tests so it's clear what the new functions do?

File: awx/connection.go [hunk 2]
```diff
@@ -533,3 +534,49 @@ func (c *Connection) indent(data []byte) []byte {
 	}
 	return buffer.Bytes()
 }
+
+var passwordFilterRegex = regexp.MustCompile("(?i)(password|token|authorization|key)")
```
Comment (moolitayer) [inline awx/connection.go]: Do we need the extra match group?
Please use `"(?i:password|token|authorization|key)"`

(I think according to the [doc](https://github.com/google/re2/wiki/Syntax) the current syntax should only effect the current group?)

File: awx/connection.go [hunk 3]
```diff
@@ -533,3 +534,49 @@ func (c *Connection) indent(data []byte) []byte {
...
+var passwordFilterRegex = regexp.MustCompile("(?i:password|token|authorization|key)")
+
+func filterJsonBytes(bytes []byte) []byte {
+	if len(bytes) == 0 {
+		return bytes
+	}
+	var jsonObj interface{}
+	err := json.Unmarshal(bytes, &jsonObj)
+	if err != nil {
+		glog.Infof("Error parsing: %v", err)
```
Comment (moolitayer) [inline awx/connection.go]: warn?

File: awx/connection.go [hunk 4]
```diff
@@ -533,3 +534,49 @@ func (c *Connection) indent(data []byte) []byte {
...
+		return []byte{}
+	}
+	jsonObj = filterJsonObject(jsonObj.(map[string]interface{}))
```
Comment (moolitayer) [inline awx/connection.go]: What if jsonObj is of a different type? (e.g the server returns a list of projects )

File: awx/conntection_test.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+		t.Errorf("Expected %s, got %s", expected, result)
+	}
+
+	input = []byte("{\"aaa\":[{\"password\":\"foo\"},\"bar\"]}")
+	expected = []byte("{\"aaa\":[{\"password\":\"REDACTED\"},\"bar\"]}")
+	result = filterJsonBytes(input)
+	if string(result) != string(expected) {
+		t.Errorf("Expected %s, got %s", expected, result)
+	}
+}
```
Comment (moolitayer) [inline awx/conntection_test.go]: Tests look great, thanks!

**Comment by moolitayer** [comment]
@zgalor @jhernand @yaacov another set of eyes please

File: awx/connection.go (hunk 4 above)
Comment (elad661) [reply awx/connection.go]: Thanks @moolitayer, this cast was a leftover from a previous version and is definitely not needed. I also added an extra test to make sure we properly supports lists as the root object.

**Review by nimrodshn — APPROVED** [review]
Very nice.
LGTM.

**Review by moolitayer — APPROVED** [review]
LGTM :+1:

**Comment by moolitayer** [comment]
Thanks @elad661 :tada:

**Pull request merged**